### PR TITLE
Laboratory work #2, Andrey Gusev - 22FPL1 

### PR DESCRIPTION
**Pull request by AndreyGusev1**

### CODE REVIEW

File: lab_2_tokenize_by_bpe/main.py [hunk 1]
```diff
@@ -26,6 +35,15 @@ def collect_frequencies(
     :param end_of_word: a token that signifies the end of word
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(text, str) or not isinstance(start_of_word, str | None) or not isinstance(end_of_word, str):
+        return None
+    dictionary = {}
+    for i in text.split():
+        if prepare_word(i, start_of_word, end_of_word) is None:
+            return None
+        else:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:44]: not need to use `else` if you return in if statement

File: lab_2_tokenize_by_bpe/main.py [hunk 2]
```diff
@@ -26,6 +35,15 @@ def collect_frequencies(
...
+            dictionary.update({prepare_word(i, start_of_word, end_of_word): text.split().count(i)})
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:45]: I think it is wrong

File: lab_2_tokenize_by_bpe/main.py [hunk 3]
```diff
@@ -36,6 +54,17 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pair_dictionary = {}
+    for word, frequency in word_frequencies.items():
+        if len(word) >= 2:
+            for i in range(len(word)-1):
+                if (word[i], word[i+1]) in pair_dictionary:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:63]: too many indents

Comment (artyomtugaryov) [reply lab_2_tokenize_by_bpe/main.py:63]: simplify

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 1 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 3 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 2 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: Can you explain a little bit, I don't understand why. All tests are green and I checked the function works correctly.

File: lab_2_tokenize_by_bpe/main.py [hunk 4]
```diff
@@ -26,6 +35,15 @@ def collect_frequencies(
...
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(text, str) or not isinstance(start_of_word, str | None) or not isinstance(end_of_word, str):
+        return None
+    dictionary = {}
+    for i in text.split():
+        symbols = prepare_word(i, start_of_word, end_of_word)
+        if symbols is None:
+            return None
+        dictionary.update({symbols: text.split().count(i)})
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:45]: You add the same value many times to the dict

File: lab_2_tokenize_by_bpe/main.py [hunk 5]
```diff
@@ -36,6 +54,16 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pair_dictionary = {}
+    for word, frequency in word_frequencies.items():
+        for i in range(len(word)-1):
+            if (word[i], word[i+1]) in pair_dictionary:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:62]: simplify - set 0 if there is not the key in the dict and do += always

File: lab_2_tokenize_by_bpe/main.py [hunk 6]
```diff
@@ -47,6 +75,23 @@ def merge_tokens(
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+    dictionary = {}
+    united = pair[0]+pair[1]
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:81]: concatination of the string? Use fstring

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 4 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 5 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 6 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py [hunk 7]
```diff
@@ -47,6 +76,21 @@ def merge_tokens(
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+    dictionary = {}
+    united = f'{pair[0]}{pair[1]}'
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:82]: just `join`

File: lab_2_tokenize_by_bpe/main.py [hunk 8]
```diff
@@ -47,6 +76,21 @@ def merge_tokens(
...
+    for symbols in word_frequencies.keys():
+        new_symbols = list(symbols)
+        i = 0
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:85]: for what?

File: lab_2_tokenize_by_bpe/main.py [hunk 9]
```diff
@@ -58,6 +102,22 @@ def train(
     :param num_merges: required number of new tokens
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(num_merges, int):
+        return None
+    nums = 0
+    while nums != num_merges:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:108]: Looks like not just `!=`, but `<`

File: lab_2_tokenize_by_bpe/target_score.txt [hunk 10]
```diff
@@ -1 +1 @@
-0
+6
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/target_score.txt:1]: only 6?(

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 7 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 8 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py (hunk 9 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/main.py [hunk 11]
```diff
@@ -58,6 +102,22 @@ def train(
...
+        pairs = count_tokens_pairs(word_frequencies)
+        if pairs is None:
+            return None
+        if pairs == {}:
```
[the remark this answers is not in the document]
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

File: lab_2_tokenize_by_bpe/target_score.txt (hunk 10 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/target_score.txt:1]: Didn't have enough time to do all the tasks and make my PR green in time, so decided to leave only 6 :(

File: lab_2_tokenize_by_bpe/main.py [hunk 12]
```diff
@@ -36,6 +56,15 @@ def count_tokens_pairs(
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pair_dictionary = {}
+    for word, frequency in word_frequencies.items():
+        for i in range(len(word)-1):
+            if (word[i], word[i+1]) not in pair_dictionary:
```
Comment (artyomtugaryov) [inline lab_2_tokenize_by_bpe/main.py:64]: please, `pair` variable

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_2_tokenize_by_bpe/main.py (hunk 12 above)
Comment (AndreyGusev1) [reply lab_2_tokenize_by_bpe/main.py]: fixed

**Pull request closed without merging**